tests/main/uc20-snap-recovery-encrypt/task.yaml

```diff
@@ -0,0 +1,120 @@
+    echo "label: gpt" | sfdisk "$LOOP"
+
+    # XXX: update to the UC20 gadget when it's available
+    echo "Get the UC16 gadget"
+    snap download pc
+    unsquashfs -d gadget-dir pc_*.snap
+
+execute: |
+    LOOP="$(cat loop.txt)"
+    echo "Run the snap-bootstrap tool"
+    /usr/lib/snapd/snap-bootstrap create-partitions ./gadget-dir "$LOOP"
+
+    echo "And check that the partitions are created"
+    sfdisk -l "$LOOP" | MATCH '1M\s+BIOS boot'
+    sfdisk -l "$LOOP" | MATCH '50M\s+EFI System'
+    # we used "lsblk --fs" here but it was unreliable
+    mkdir -p ./mnt
+    mount "${LOOP}p2" ./mnt
+    df -T "${LOOP}p2" | MATCH vfat
+    umount ./mnt
+    file -s "${LOOP}p2" | MATCH 'label: "system-boot"'
+
+    echo "Check that the non-fs content was deployed"
+    dd if="${LOOP}p1" of=bios-boot.img
+    # the bios-boot.img is 1Mb
+    test "$(stat --printf="%s" ./bios-boot.img)" = 1048576
+    # we truncate it to the size of pc-core.img
+    truncate --size="$(stat --printf="%s" ./gadget-dir/pc-core.img)" bios-boot.img
+    # and verify that its the same
+    cmp bios-boot.img ./gadget-dir/pc-core.img
+
+    echo "Check that the filesystem was not auto-mounted"
+    mount
+    mount | not MATCH /run/mnt/system-boot
+
+    echo "Check that the filesystem content was deployed"
+    mkdir -p ./mnt
+    mount "${LOOP}p2" ./mnt
+    ls ./mnt/EFI/boot/grubx64.efi
+    ls ./mnt/EFI/boot/bootx64.efi
+    ls ./mnt/EFI/ubuntu/grub.cfg
+    umount ./mnt
+
+    # debug message to see if the udev database is correctly updated
+    udevadm info --query=property "${LOOP}p2" | grep ID_FS_TYPE ||:
+
+    echo "now add a partition"
+    cat >> gadget-dir/meta/gadget.yaml <<EOF
+          - name: Other ext4
+            type: 83,0FC63DAF-8483-4772-8E79-3D69D8477DE4
+            filesystem: ext4
+            filesystem-label: other-ext4
+            size: 110M
+          - name: Boot partition
+            type: 83,0FC63DAF-8483-4772-8E79-3D69D8477DE4
+            role: system-boot
+            filesystem: ext4
+            size: 120M
+          - name: Data partition
+            type: 83,0FC63DAF-8483-4772-8E79-3D69D8477DE4
+            role: system-data
+            filesystem: ext4
+            size: 130M
+    EOF
+    /usr/lib/snapd/snap-bootstrap create-partitions --encrypt --key-file=keyfile --mount ./gadget-dir "$LOOP"
+    sfdisk -l "$LOOP" | MATCH '110M\s* Linux filesystem'
+    sfdisk -l "$LOOP" | MATCH '120M\s* Linux filesystem'
+    sfdisk -l "$LOOP" | MATCH '130M\s* Linux filesystem'
+
+    echo "check that the filesystems are created and mounted"
+    mount
+
+    mount | MATCH /run/mnt/other-ext4
+    df -T "${LOOP}p3" | MATCH ext4
+    file -s "${LOOP}p3" | MATCH 'volume name "other-ext4"'
+    umount /run/mnt/other-ext4
+
+    mount | MATCH /run/mnt/ubuntu-boot
+    df -T "${LOOP}p4" | MATCH ext4
+    file -s "${LOOP}p4" | MATCH 'volume name "ubuntu-boot"'
+    umount /run/mnt/ubuntu-boot
+
+    cryptsetup isLuks "${LOOP}p5"
```

Comment:
Should we have something like `not cryptsetup isLuks "${LOOP}p4` (and for p3) as well? To double check that the system-role is honored when applying encryption?